src/memory.rs

```diff
@@ -0,0 +1,235 @@
+            self.seek(SeekFrom::Current(chunk.len() as i64))?;
+            total_bytes_written += chunk.len();
+        }
+        Ok(total_bytes_written)
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        // nothing to do
+        Ok(())
+    }
+}
+
+impl Seek for Memory {
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64> {
+        self.addr_seek.seek(pos)
+    }
+}
+
+pub struct PaddedMemory {
+    drv: Rc<RefCell<Box<dyn Introspectable>>>,
+    addr_seek: AddressSeek,
+}
+
+impl PaddedMemory {
+    pub fn new(drv: Rc<RefCell<Box<dyn Introspectable>>>) -> StdResult<Self, Box<dyn StdError>> {
+        Ok(PaddedMemory {
+            drv: drv.clone(),
+            addr_seek: AddressSeek {
+                pos: 0,
+                max_addr: drv.borrow().get_max_physical_addr()?,
+            },
+        })
+    }
+}
+
+// TODO: slight duplication
+impl Read for PaddedMemory {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        // amount of bytes we need to read
+        let mut read_remain: usize = buf.len();
+        let mut bytes_read: usize = 0;
+        while read_remain > 0 {
+            // determine size of next chunk
+            let paddr = self.stream_position()?;
+            let frame = PageFrame::with_paddr(paddr);
+            // windows_len -> 4K or less, if offset in frame
+            let next_chunk_size = std::cmp::min(frame.window_len(), read_remain as u32) as usize;
+            let chunk_end = bytes_read + next_chunk_size;
+            // get chunk
+            let chunk = &mut buf[bytes_read..chunk_end];
+            // read the frame
+            match self.drv.borrow().read_frame(frame, chunk) {
+                // handle non existing frames by padding
+                Err(ref e) if e.kind() == ErrorKind::NotFound => {
+                    trace!("PaddedMemory: frame not found: {:X}", frame.number);
+                    chunk.fill(0)
+                }
+                Err(e) => return Err(e),
+                _ => (),
+            };
+            // advance pos
+            self.seek(SeekFrom::Current(next_chunk_size as i64))?;
+            // update loop vars
+            bytes_read += next_chunk_size;
+            read_remain -= next_chunk_size;
+        }
+        Ok(bytes_read as usize)
+    }
+}
+
+impl Seek for PaddedMemory {
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64> {
+        self.addr_seek.seek(pos)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Seek
+    #[test]
+    fn test_seek_start() -> Result<()> {
+        let max_addr: u64 = 1000;
+        let mut mock_introspectable = MockIntrospectable::new();
+        mock_introspectable
+            .expect_get_max_physical_addr()
+            .returning(move || Ok(max_addr));
+        let mut memory = Memory::new(Rc::new(RefCell::new(Box::new(mock_introspectable)))).unwrap();
+        // seek 0 doesn't move position
+        assert_eq!(0, memory.seek(SeekFrom::Start(0))?);
```

Comment:
Don't put multiple unrelated asserts into the same test case, if you want to avoid code duplication use the `test-case` crate. See https://github.com/Wenzel/libmicrovmi/blob/master/src/driver/kvm.rs#L398